File: 010_Regular_Expression_Matching/solution.py
class Solution(object):
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        if not s:
            return not p or (len(p) > 1 and p[1] == '*' and self.isMatch(s,p[2:]))
        elif not p:
            return False
        
        if len(p) == 1:
            return len(s) == 1 and (p[0] == '.' or p[0] == s[0])
        
        if p[1] != '*':
            return (p[0] == '.' or p[0] == s[0]) and self.isMatch(s[1:], p[1:])
        else:
            if (s[0] != p[0] and p[0] != '.'):
                return self.isMatch(s, p[2:])
            else:
                if self.isMatch(s, p[2:]):
                    return True
                    
                i = 0;
                while i < len(s) and (s[i] == p[0] or p[0] == '.'):
                    if self.isMatch(s[i+1:], p[2:]):
                        return True
                    i += 1

                return False

File: 010_Regular_Expression_Matching/solution.go
func isMatch(s string, p string) bool {
    lens := len(s)
    lenp := len(p)
    matched := make([][]bool, lens + 1)
    for i := 0; i <= lens; i ++ {
        matched[i] = make([]bool, lenp + 1)
    }
    matched[0][0] = true
    for i := 0; i <= lens; i ++ {
        for j := 1; j <= lenp; j ++ {
            if j > 1 && p[j - 1] == '*' {
                matched[i][j] = matched[i][j - 2] || (i > 0 && (s[i - 1] == p[j - 2] || p[j - 2] == '.') && matched[i - 1][j])
            } else {
                matched[i][j] = i > 0 && matched[i - 1][j - 1] && (s[i - 1] == p[j - 1] || p[j - 1] == '.')
            }
        }
    }
    return matched[lens][lenp]
}
